### PR TITLE
LibWeb+LibGfx: The Border Radii Trilogy: Book Three: Box-Shadows

### PR DESCRIPTION
**Pull request by MacDue**
|  LibWeb                                                                                                         | Chromium                                                                                                        |
|-----------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------|
| image | image |

_It's not perfect but it's good enough for me :^)_

---

There is a bit of a performance hit depending on how big the element is and how big the radius is (bigger = worse).
This is because the larger the border radii are the bigger the corner bitmap needs to be, so more that has to be blurred by `FastBoxBlurFilter` (where most of the time is spent). I've done some obvious little improvements to `FastBoxBlurFilter`, but there's probably much more that could be done (5000 points to someone who figures how how to turn `FastBoxBlurFilter` to 
`RapidBoxBlurFilter` -- or figures out how to blur less, there's some symmetry + wasted time blurring infill).

This is still not much of an issue on most sites that use box-shadow rather sparingly.  

---

`car.png` is CC0, took it myself :^)

### CODE REVIEW

File: Userland/Libraries/LibWeb/Painting/ShadowPainting.cpp [hunk 1]
```diff
@@ -38,79 +54,214 @@ void paint_box_shadow(PaintContext& context, Gfx::IntRect const& content_rect, V
                 painter.fill_rect(rect, color);
         };
 
-        // If there's no blurring, we can save a lot of effort.
-        if (box_shadow_data.blur_radius == 0) {
+        // If there's no blurring, or corners, we can save a lot of effort.
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Painting/ShadowPainting.cpp:57]:
```suggestion
        // If there's no blurring, nor rounded corners, we can save a lot of effort.
```

File: Userland/Libraries/LibWeb/Painting/ShadowPainting.cpp [hunk 2]
```diff
@@ -39,79 +54,214 @@ void paint_box_shadow(PaintContext& context, Gfx::IntRect const& content_rect, B
...
             return;
         }
         auto shadow_bitmap = shadows_bitmap.release_value();
         Gfx::Painter corner_painter { *shadow_bitmap };
-        auto double_radius = box_shadow_data.blur_radius * 2;
-        corner_painter.fill_rect(all_corners_rect.shrunken(double_radius, double_radius, double_radius, double_radius), box_shadow_data.color);
+        Gfx::AntiAliasingPainter aa_corner_painter { corner_painter };
+
+        aa_corner_painter.fill_rect_with_rounded_corners(shadow_bitmap_rect.shrunken(double_radius, double_radius, double_radius, double_radius), box_shadow_data.color, top_left_shadow_corner, top_right_shadow_corner, bottom_right_shadow_corner, bottom_left_shadow_corner);
+        // FIXME: Make fast box blur faster
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Painting/ShadowPainting.cpp:149]: :yakspeed:

File: Userland/Libraries/LibWeb/Painting/ShadowPainting.cpp [hunk 3]
```diff
@@ -39,79 +54,214 @@ void paint_box_shadow(PaintContext& context, Gfx::IntRect const& content_rect, B
...
+
+        auto left_start = inner_bounding_rect.left() - blurred_edge_thickness;
+        auto right_start = inner_bounding_rect.left() + inner_bounding_rect.width();
+        auto top_start = inner_bounding_rect.top() - blurred_edge_thickness;
+        auto bottom_start = inner_bounding_rect.top() + inner_bounding_rect.height();
+
+        auto top_left_corner_blit_pos = inner_bounding_rect.top_left().translated(-blurred_edge_thickness, -blurred_edge_thickness);
+        auto top_right_corner_blit_pos = inner_bounding_rect.top_right().translated(-top_right_corner_size.width() + 1 + double_radius, -blurred_edge_thickness);
+        auto bottom_left_corner_blit_pos = inner_bounding_rect.bottom_left().translated(-blurred_edge_thickness, -bottom_left_corner_size.height() + 1 + double_radius);
+        auto bottom_right_corner_blit_pos = inner_bounding_rect.bottom_right().translated(-bottom_right_corner_size.width() + 1 + double_radius, -bottom_right_corner_size.height() + 1 + double_radius);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Painting/ShadowPainting.cpp:212]: The `1`s could use a comment, or variable

File: Userland/Libraries/LibWeb/Painting/ShadowPainting.cpp [hunk 4]
```diff
@@ -39,79 +54,214 @@ void paint_box_shadow(PaintContext& context, Gfx::IntRect const& content_rect, B
...
+        // │xxx│xxxxxxxx│xxx│
+        // │xxx│xxxxxxxx│xxx│
+        // │xxx│xxxxxxxx│xxx│
+        // │xxx│xxxxxxxx│xxx│
+        // ├───┼────────┼───┤
+        // │   │ xxxxxx │   │
+        // └───┴────────┴───┘
+
+        // How many times would you like to paint the shadow sir?
+        // Yes.
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Painting/ShadowPainting.cpp:261]: :laughing:

**Review by linusg — COMMENTED** [review]
Awesome! A few remarks:

- First commit needs an explanation please, makes no sense out of context
- Please avoid @ mentioning people in commits - it'll send them a notification whenever someone rebases & pushes to their fork. just `Lubrsi` instead of `@Lubrsi` is fine. Tag them in the PR if you'd like them to know :)

**Comment by MacDue** [comment]
Updated to add 'Fixes #14325' (didn't notice an issue had been created for this).

**Pull request merged**